pkg/server/endpoints/bundle/acme_auth.go

```diff
@@ -0,0 +1,192 @@
+package bundle
+
+import (
+	"context"
+	"crypto"
+	"crypto/ecdsa"
+	"crypto/tls"
+	"crypto/x509"
+	"encoding/asn1"
+	"io"
+	"math/big"
+
+	"github.com/sirupsen/logrus"
+	"github.com/spiffe/spire/pkg/common/cryptoutil"
+	"github.com/spiffe/spire/pkg/common/version"
+	"github.com/spiffe/spire/pkg/server/endpoints/bundle/internal/autocert"
+	"github.com/spiffe/spire/proto/spire/server/keymanager"
+	"github.com/zeebo/errs"
+	"golang.org/x/crypto/acme"
+)
+
+const (
+	acmeKeyPrefix = "bundle-acme-"
+)
+
+// ACMECache implements a cache for the autocert manager. It makes some
+// simplifying assumptions based on our usage for the bundle endpoint. Namely,
+// it assumes there is going to be a single cache entry, since we only support
+// a single domain. It assumes PEM encoded blocks of data and strips out the
+// private key to be stored in the key manager instead of on disk with the rest
+// of the data.
+type ACMEConfig struct {
+	// DirectoryURL is the ACME directory URL
+	DirectoryURL string
+
+	// DomainName is the domain name of the certificate to obtain.
+	DomainName string
+
+	// CacheDir is the directory on disk where we cache certificates.
+	CacheDir string
+
+	// Email is the email address of the account to register with ACME
+	Email string
+}
+
+func ACMEAuth(log logrus.FieldLogger, km keymanager.KeyManager, config ACMEConfig) ServerAuth {
+	// The acme client already defaulting to Let's Encrypt if the URL is unset
+	// but we want it populated for logging purposes.
+	if config.DirectoryURL == "" {
+		config.DirectoryURL = acme.LetsEncryptURL
+	}
+
+	return &acmeAuth{
+		m: &autocert.Manager{
+			Prompt: func(tosURL string) bool {
+				log.WithFields(logrus.Fields{
+					"directory_url": config.DirectoryURL,
+					"tos_url":       tosURL,
+					"email":         config.Email,
+				}).Info("ACME Terms of Service accepted")
+				return true
+			},
+			Email:      config.Email,
+			Cache:      autocert.DirCache(config.CacheDir),
+			HostPolicy: autocert.HostWhitelist(config.DomainName),
+			Client: &acme.Client{
+				DirectoryURL: config.DirectoryURL,
+				UserAgent:    "SPIRE-" + version.Version(),
+			},
+			KeyStore: &acmeKeyStore{
+				log: log,
+				km:  km,
+			},
+		},
+	}
+}
+
+type acmeAuth struct {
+	m *autocert.Manager
+}
+
+func (a *acmeAuth) GetTLSConfig() *tls.Config {
+	return a.m.TLSConfig()
+}
+
+type acmeKeyStore struct {
+	log logrus.FieldLogger
+	km  keymanager.KeyManager
+}
+
+func (ks *acmeKeyStore) GetPrivateKey(ctx context.Context, id string) (crypto.Signer, error) {
+	keyID := acmeKeyPrefix + id
```

Comment:
Since we limit ourselves to a single domain, the upper limit is three. 
* account key
* ecdsa key for domain
* rsa key for domain

If we want to deny clients that only support one protocol or the other, we can get it down to two.